File: 20-testing/power.py
import pytest

def power(base, exponent):
    assert(exponent >= 0)
    if exponent == 0:
         return 1
    if exponent == 1:
         return base
    if exponent & 1 == 1:
         return base * power(base*base, (exponent-1)//2)

class TestPower:
    def test_knowngood(self):
        assert power(4,7) == 16384

    def test_knownerror(self):
        with pytest.raises(AssertionError):
            power(2,-3)

File: 06-testing/power/power.js
export function power(base, exponent) { 
  let answer = 1;
  while (exponent > 0) {
    if (exponent & 1) {
      answer = answer * base;
      exponent--;
    }
    base = base * base;
    exponent = exponent / 2;
  }
  return answer;
}
